# **Análisis 2: Análisis de resultados diferenciando materias de modalidad**

Es oportuno hacer un análisis de los datos teniendo en cuenta las distintas materias de modalidad.

**Objetivo:**

* Identificar las materias de modalidad: Filtrar el DataFrame para incluir solo las columnas relevantes para las materias de modalidad.

* Agrupar por materia de modalidad: Agrupar los datos por la columna 'Materia modalidade'.

* Análisis estadístico por materia: Calcular los estadísticos descriptivos (media, mediana, desviación estándar, etc.) para la columna 'Cualif. Materia Modalidade' dentro de cada grupo (cada materia de modalidad).

* Visualización por materia: Generar histogramas y/o box plots de la columna 'Cualif. Materia Modalidade' para cada materia de modalidad para visualizar sus distribuciones.

* Comparar y concluir: Analizar los resultados estadísticos y las visualizaciones para comparar el rendimiento en las diferentes materias de modalidad e identificar posibles diferencias.


## 1. Identificar las materias de modalidad

Filtramos el DataFrame para incluir solo las columnas relevantes para las materias de modalidad, las cuales son: ´Materia de modalidade´ y ´Cualif. Materia de Modalidade´


In [9]:
import pandas as pd

df = pd.read_excel("proba_xeral.xlsx")

modalidad_df = df[['Apelidos','Nome', 'Nom. Materia modalidade', 'Materia Modalidade']].copy()

display(modalidad_df)


Apelidos           Nome     Nom. Materia modalidade  \
0                ÁLVAREZ GRAÑA           ERIC          Artes Escénicas II   
1               AMMI GUAROUACH           ZARA                    Latín II   
2                ARIAS PORTELA           JOSÉ              Matemáticas II   
3               BARREIRO ALDAO          LIDIA         Análisis Musical II   
4                   CABO SENRA         MARTÍN              Matemáticas II   
5                  CALVO DURÁN           ADÁN              Matemáticas II   
6               CASADO MIRANDA           UXÍA          Artes Escénicas II   
7              CASTRO GONZÁLEZ         ESTELA              Matemáticas II   
8           FARAMIÑÁN FERVENZA           HUGO         Debuxo Artístico II   
9              FEIJOÓ GONZÁLEZ           HUGO              Matemáticas II   
10              FREIJEIRO ARCA           ALBA          Artes Escénicas II   
11               GARCÍA BARROS          NEREA         Debuxo Artístico II   
12                GARCÍA LORES         SANDRA  Matemáticas Apl. ás CC.SS.   
13             GARCÍA MARTÍNEZ   ACRACIA SARA         Debuxo Artístico II   
14                 GÓMEZ NOVAS        CECILIA         Debuxo Artístico II   
15             HOLGADO PIÑEIRO            IVO          Artes Escénicas II   
16            IGLESIAS MONTERO          CELIA         Debuxo Artístico II   
17                 LAMAS PAZOS           EMMA  Matemáticas Apl. ás CC.SS.   
18                 LEÓN MESÍAS            ANA         Debuxo Artístico II   
19               LOURO ESPERÓN          DAVID  Matemáticas Apl. ás CC.SS.   
20              MAIT RODRÍGUEZ          ZAIRA         Debuxo Artístico II   
21              MARTÍNEZ PAZOS        CLAUDIA  Matemáticas Apl. ás CC.SS.   
22  MENDONÇA MIRANDA HENRIQUES          LUNNA  Matemáticas Apl. ás CC.SS.   
23               MURO HILLERUD  DELIA NANDINI  Matemáticas Apl. ás CC.SS.   
24               OLTRA LOURIDO        CANDELA         Debuxo Artístico II   
25                 OTERO DURÁN         ADRIÁN  Matemáticas Apl. ás CC.SS.   
26             PAZOS DOMÍNGUEZ      ALEJANDRO  Matemáticas Apl. ás CC.SS.   
27                PEÓN GÁNDARA           IAGO              Matemáticas II   
28                 PÉREZ PENAS            MAX         Debuxo Artístico II   
29                  PILAY RUIZ    ELVIS BRYAN  Matemáticas Apl. ás CC.SS.   
30             QUINTANA GARCÍA          MARIO  Matemáticas Apl. ás CC.SS.   
31            RODRÍGUEZ AGUIAR           SARA                    Latín II   

    Materia Modalidade  
0                 4.50  
1                 7.00  
2                 8.75  
3                 6.50  
4                 9.50  
5                 7.75  
6                 5.00  
7                 8.25  
8                 7.00  
9                 5.75  
10                5.00  
11                5.00  
12                7.25  
13                9.25  
14                6.00  
15                5.00  
16                6.00  
17                7.50  
18                8.00  
19                9.00  
20                6.25  
21                6.00  
22                6.25  
23                5.00  
24                6.00  
25                8.25  
26                9.50  
27                8.50  
28                9.00  
29                3.50  
30                7.25  
31                5.50

In [11]:
import matplotlib.pyplot as plt
import seaborn as sns


plt.figure(figsize=(8, 5))
ax = sns.barplot(x=modalidad_counts.index, y=modalidad_counts.values, palette='Set3')
plt.title('Número de Estudiantes con Calificación por Materia de Modalidad')
# plt.xlabel('Modality Subject')
plt.ylabel('Number of Students')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()

# Add numerical values on top of each bar
for p in ax.patches:
    ax.annotate(f'{int(p.get_height())}', (p.get_x() + p.get_width() / 2., p.get_height()),
                ha='center', va='center', xytext=(0, 5), textcoords='offset points')

plt.show()


NameError: name 'modalidad_counts' is not defined

In [ ]:

modalidad_df.groupby('Nom. Materia modalidade')['Materia Modalidade'].describe().transpose()


## 2. Agrupar por materia de modalidad


Agrupar los datos por la columna 'Materia modalidade' para preparar el análisis estadístico y la visualización por materia.


In [6]:
modalidad_stats = modalidad_df['Materia Modalidade'].describe().to_frame().T
display(modalidad_stats)


count     mean       std  min     25%   50%   75%  max
Materia Modalidade   32.0  6.84375  1.634741  3.5  5.6875  6.75  8.25  9.5

## 2. Visualización por materia

Generar histogramas y box plots de la columna 'Materia Modalidade' para cada materia de modalidad para visualizar sus distribuciones.


In [7]:
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import math

# Get unique modalities, sort alphabetically, and create a color mapping using 'Set3'
modalities = sorted(modalidad_df['Nom. Materia modalidade'].unique())
colors = sns.color_palette('Set3', len(modalities))
color_map = dict(zip(modalities, colors))

# Histograms
num_modalidades = len(modalities)
cols = 3
rows = math.ceil(num_modalidades / cols)

plt.figure(figsize=(cols * 5, rows * 4))

for i, modality in enumerate(modalities):
    plt.subplot(rows, cols, i + 1)
    group = grouped_modalidad_df.get_group(modality)
    plt.hist(group, bins=10, color=color_map[modality]) # Use color map
    plt.title(f'Distribución de Cualificacións para {modality}')
    plt.xlabel('Cualificación')
    plt.ylabel('Frecuencia')
    plt.xlim(0, 10)
    plt.ylim(0, 4) # Set fixed y-axis limit
    plt.yticks(np.arange(0, 5, 1)) # Set fixed y-ticks

    plt.grid(axis='y', linestyle='--')

plt.tight_layout()
plt.show()

# Box plot with individual points and means
plt.figure(figsize=(10, 6))
sns.boxplot(data=modalidad_df, x='Nom. Materia modalidade', y='Materia Modalidade', hue='Nom. Materia modalidade', palette=color_map, order=modalities, legend=False) # Use color map and specify order
sns.swarmplot(data=modalidad_df, x='Nom. Materia modalidade', y='Materia Modalidade', color='black', size=5, ax=plt.gca(), order=modalities) # Add individual points and specify order

# Add mean points for each modality
modalities_mean = grouped_modalidad_df.mean().reindex(modalities) # Reindex means to match sorted order
plt.scatter(x=range(len(modalities)), y=modalities_mean.values, color='white', edgecolors='black', s=100, zorder=3) # Add white points with black edge

plt.title('Box Plot de Cualificacións por Materia de Modalidade con Puntos Individuais e Medias')
plt.xlabel('Materia de Modalidade')
plt.ylabel('Cualificación')
plt.ylim(0, 10)
plt.yticks(np.arange(0, 11, 1)) # Set y-ticks from 0 to 10 with step 1
plt.xticks(rotation=45, ha='right')

# Add horizontal grid lines
plt.grid(axis='y', linestyle='--')



plt.legend()
plt.tight_layout()
plt.show()


NameError: name 'grouped_modalidad_df' is not defined

## 3. Conclusiones

Analizamos los resultados estadísticos y las visualizaciones para comparar el rendimiento en las diferentes materias de modalidad e identificar posibles diferencias.


### 3.1. **Tendencias generales del rendimiento:**

* La **media global** de las materias de modalidad es **6,84**, y la **mediana global** es **6,75**.
  → Esto indica un rendimiento medio-alto en conjunto, con ligeros sesgos positivos (la media está ligeramente por encima de la mediana).

* La mayoría de las asignaturas parecen tener pocos alumnos, lo cual es habitual en materias de 2º de Bachillerato. Las calificaciones tienden a concentrarse entre el 5 y el 8, reflejando en general un nivel medio de rendimiento. Las materias con más alumnos permiten una mejor observación de la distribución y muestran una mayor diversidad de resultados.

### 3.2. **Análisis por materia de modalidad**

#### **1. Análisis Musical II**  
- **Datos estadísticos**:  
  - Solo hay un dato (count = 1.0) con una calificación de **6.5**.  
  - No hay desviación estándar (std = NaN) porque solo hay un valor.  
- **Histograma**:  
  - No se puede analizar por falta de datos suficientes.  
- **Box Plot**:  
  - Se representa como una línea única en **6.5** (sin rango intercuartílico ni bigotes).  
- **Conclusión**:  
  - Muestra insuficiente para análisis significativo. Podría tratarse de un error o una materia con muy pocos estudiantes.  

---

#### **2. Artes Escénicas II**  
- **Datos estadísticos**:  
  - **4 estudiantes**, media = **4.875**, desviación estándar baja (**0.250**).  
  - Rango estrecho: **4.5 a 5.0** (sin outliers).  
- **Histograma**:  
  - Concentrado en **5.0**, con una pequeña frecuencia en **4.5**.  
- **Box Plot**:  
  - La caja es estrecha (Q1 = 4.875, Q3 = 5.0), confirmando poca variabilidad.  
  - No hay bigotes significativos, ya que todos los datos están muy agrupados.  
- **Conclusión**:  
  - **Rendimiento homogéneo y alto**. Posible evaluación poco discriminante o grupo muy uniforme.  

---

#### **3. Debuxo Artístico II**  
- **Datos estadísticos**:  
  - **9 estudiantes**, media = **6.944**, desviación estándar moderada (**1.483**).  
  - Rango amplio: **5.0 a 9.25**.  
- **Histograma**:  
  - Distribución ligeramente sesgada a la izquierda, con mayoría en **6.0–8.0** y un caso bajo (**5.0**).  
- **Box Plot**:  
  - Caja muestra Q1 = 6.0, Q3 = 8.0 (rango intercuartílico amplio).  
  - Bigotes llegan a **5.0** (mínimo) y **9.25** (máximo).  
  - Posible outlier en **5.0** (punto alejado de la caja).  
- **Conclusión**:  
  - **Mayoría con buen rendimiento, pero un estudiante rezagado**. Podría requerir apoyo individualizado.  

---

#### **4. Matemáticas Apl. ás CC.SS.**  
- **Datos estadísticos**:  
  - **10 estudiantes**, media = **6.95**, alta desviación estándar (**1.825**).  
  - Rango muy amplio: **3.5 a 9.5**.  
- **Histograma**:  
  - Distribución bimodal o con outliers (picos en extremos bajos y altos).  
- **Box Plot**:  
  - Caja centrada en **6.06–8.06** (Q1–Q3), pero con bigotes hasta **3.5** y **9.5**.  
  - **Outlier claro en 3.5** (punto fuera del bigote inferior).  
- **Conclusión**:  
  - **Gran variabilidad**. Algunos estudiantes sobresalen (9.5), pero otros fracasan (3.5). Sugiere diferencias en comprensión o metodología de enseñanza.  

---

#### **5. Matemáticas II**  
- **Datos estadísticos**:  
  - **6 estudiantes**, media = **8.083**, desviación estándar moderada (**1.281**).  
  - Rango: **5.75 a 9.5**.  
- **Histograma**:  
  - Sesgo a la derecha, con mayoría en **7.5–9.5** y un caso bajo (**5.75**).  
- **Box Plot**:  
  - Caja estrecha (Q1 = 7.875, Q3 = 8.687), pero con bigote largo hacia abajo (**5.75**).  
  - **Outlier en 5.75** (alejado del grupo principal).  
- **Conclusión**:  
  - **Excelente rendimiento general, pero un estudiante con dificultades**. Podría ser un caso aislado que necesita refuerzo.  

---

### 3.3. **Conclusión General**  
1. **Materias artísticas (Artes Escénicas II, Debuxo Artístico II)**:  
   - Resultados más homogéneos, especialmente en Artes Escénicas II.  
   - En Debuxo, hay un posible outlier que sugiere atención individual.  

2. **Matemáticas (Matemáticas Apl. ás CC.SS. y Matemáticas II)**:  
   - **Matemáticas Apl. ás CC.SS.** muestra **gran dispersión**, con estudiantes en extremos opuestos (éxito vs. fracaso).  
   - **Matemáticas II** tiene un rendimiento alto en general, pero con un outlier bajo.  

3. **Análisis Musical II**:  
   - Datos insuficientes para análisis.  

### 3.4. **Recomendaciones**

- **Matemáticas Apl. ás CC.SS.**: Investigar causas de la dispersión (¿dificultad de la materia? ¿métodos de evaluación?).  
- **Matemáticas II y Debuxo Artístico II**: Revisar el caso del estudiante con calificación baja para ofrecer apoyo.  
- **Artes Escénicas II**: Evaluar si la falta de variabilidad se debe a una evaluación poco exigente o a un grupo homogéneo.  

### 3.5. **Patrón general**
- Las materias técnicas (matemáticas) presentan mayor desigualdad en resultados, mientras que las artísticas son más consistentes. Esto podría reflejar diferencias en la forma de evaluación o en la preparación previa de los estudiantes.